execution_engine/src/storage/trie_store/operations/mod.rs

```diff
@@ -252,16 +254,45 @@ where
         if !visited.insert(trie_key) {
             continue;
         }
-        let maybe_retrieved_trie: Option<Trie<K, V>> = store.get(txn, &trie_key)?;
-        match maybe_retrieved_trie {
-            // If we can't find the trie_key; it is missing and we'll return it
+        let maybe_retrieved_trie_bytes = store.get_raw(txn, &trie_key)?;
+        // Optimization: Don't look for descendants of leaves.
+        match maybe_retrieved_trie_bytes.and_then(|bytes| bytes.get(0)) {
+            // if this is a leaf don't parse (ie, the first byte is 0), just continue
+            Some(&Trie::<K, V>::LEAF_TAG) => continue,
+            // Either no entry under trie key or data is empty.  If data is empty then the trie is
+            // corrupted. Either way treat as missing.
             None => {
                 missing_descendants.push(trie_key);
+                continue;
+            }
+            _ => (),
+        }
+
+        // Parse the trie, handling errors gracefully.
+        let retrieved_trie: Trie<K, V> = match maybe_retrieved_trie_bytes
+            .map(bytesrepr::deserialize_from_slice)
+            .transpose()
+        {
+            Ok(Some(retrieved_trie)) => retrieved_trie,
+            Ok(None) => {
```

Comment:
That case is redundant now.
Maybe before `bytes.get(0)` we should unwrap or `continue`?

```rust
let retrieved_trie_bytes = match maybe_retrieved_trie_bytes {
    Some(bytes) => bytes,
    None => {
        missing_descendants.push(trie_key);
        continue;
    }
};
```

Reply:
I tweaked this logic in [9fcb768](https://github.com/casper-network/casper-node/pull/2844/commits/9fcb76807c46450853f6ca6d04e054031b4426e3) based on this suggestion.

We now don't do any special handling for the case where the returned bytes are empty, instead letting it fall through to the `Err` case when deserializing.  This should not change the net result, but will give an appropriate error log message in this case now too.